code/modules/mob/identity.dm

```diff
@@ -0,0 +1,578 @@
+	else
+		prints_list -= print
+	update_idman = TRUE
+
+/datum/mind/proc/set_print_manual(print, manual_name, category=CATEGORY_VOICEPRINTS)
+	if(print && manual_name)
+		var/list/print_entry = get_print_entry(print, category)
+		if(!print_entry)
+			switch(category)
+				if(CATEGORY_VOICEPRINTS)
+					print_entry = new(5)
+				if(CATEGORY_FACEPRINTS)
+					print_entry = new(4)
+		print_entry[1] = IDENTITY_MANUAL
+		print_entry[2] = world.time
+		print_entry[3] = manual_name
+		var/linked_print = print_entry[4]
+		if(linked_print)
+			var/linked_category = inverse_category(category)
+			var/linked_print_entry = get_print_entry(linked_print, linked_category)
+			if(linked_print_entry && linked_print_entry[4] == print)
+				linked_print_entry[1] = IDENTITY_MANUAL
+				linked_print_entry[2] = world.time
+				linked_print_entry[3] = manual_name
+				set_print_entry(linked_print, linked_print_entry, linked_category)
+		set_print_entry(print, print_entry, category)
+
+/datum/mind/proc/handle_faceprint_caching(atom/movable/seen, faceprint)
+	var/list/cache_entry = identity_cache[seen]
+	if(!cache_entry)
+		cache_entry = new(6)
+	cache_entry[3] = faceprint
+	cache_entry[4] = world.time
+	identity_cache[seen] = cache_entry
+
+/datum/mind/proc/open_idman()
+	if(!current)
+		return
+	if(!idman)
+		idman = new(src)
+	idman.ui_interact(current)
+
+/mob/living/carbon/human/default_identity_heard()
+	. = "Unknown"
+
+/mob/living/carbon/human/default_identity_seen(default_to_heard=TRUE)
+	if(w_uniform && !(wear_suit && (wear_suit.flags_inv & HIDEJUMPSUIT)) && w_uniform.identity_name)
+		. = w_uniform.identity_name
+	else if(wear_suit && wear_suit.identity_name)
+		. = wear_suit.identity_name
+	else if(default_to_heard)
+		. = default_identity_heard()
+
+/mob/living/carbon/human/default_identity_interact()
+	. = get_id_name("")
+	if(!.)
+		var/hair_text
+		var/seen_text = default_identity_seen(FALSE)
+		var/accessory_text
+		// eyecolor, haircolor+hairstyle, hat or mask, jobname
+		if(!hair_covered() && hair_color && hair_style)
+			var/htext_color = consonants(color_hex2color(hair_color))
+			var/htext_style = consonants(hair_style)
+			var/datum/sprite_accessory/hair/tortoise = hair_styles_list[hair_style]
+			if(htext_color && tortoise && tortoise.icon_state)
+				hair_text = "[htext_color] [htext_style]"
+			else
+				hair_text = "[htext_style]"
+		if(head && head.name)
+			accessory_text = " [consonants(head.name)]"
+		else if(wear_mask && wear_mask.name)
+			accessory_text = " [consonants(wear_mask.name)]"
+		. = "[hair_text][seen_text ? " [seen_text]" : ""][accessory_text]"
+
+/mob/living/carbon/human/get_faceprint()
+	if(dna)
+		. = dna.uni_identity
+
+/datum/identity_manager
+	var/cat = CATEGORY_VOICEPRINTS
+	var/datum/mind/mind
+	var/list/print_refs = list()
+	var/list/refs_lookup = list()
+	var/selected_ref
+	var/select_mode
+
+/datum/identity_manager/New(datum/mind/the_mind)
+	if(the_mind && istype(the_mind))
+		mind = the_mind
+	else
+		qdel(src)
```

Comment:
I don't this is GC safe. Regardless of what `Destroy` does, the qdeleted datum will be assigned somewhere. Unless you are just creating it with `;new /datum/identity_manager(x);`. Anything that creates one of these will have to check `if(QDELETED(x))`. Instead, you should have New CRASH if there is no mind datum and have callers assert that for you